x-pack/plugins/security/server/authentication/authenticator.test.ts

```diff
@@ -0,0 +1,655 @@
+        'Request should be a valid "KibanaRequest" instance, was [undefined].'
+      );
+    });
+
+    it('fails if login attempt is not provided.', async () => {
+      await expect(authenticator.login(requestFixture(), undefined as any)).rejects.toThrowError(
+        'Login attempt should be an object with non-empty "provider" property.'
+      );
+
+      await expect(authenticator.login(requestFixture(), {} as any)).rejects.toThrowError(
+        'Login attempt should be an object with non-empty "provider" property.'
+      );
+    });
+
+    it('fails if an authentication provider fails.', async () => {
+      const request = requestFixture();
+      const failureReason = new Error('Not Authorized');
+
+      mockBasicAuthenticationProvider.login.mockResolvedValue(
+        AuthenticationResult.failed(failureReason)
+      );
+
+      const authenticationResult = await authenticator.login(request, {
+        provider: 'basic',
+        value: {},
+      });
+      expect(authenticationResult.failed()).toBe(true);
+      expect(authenticationResult.error).toBe(failureReason);
+    });
+
+    it('returns user that authentication provider returns.', async () => {
+      const request = requestFixture();
+
+      const user = mockAuthenticatedUser();
+      mockBasicAuthenticationProvider.login.mockResolvedValue(
+        AuthenticationResult.succeeded(user, { authHeaders: { authorization: 'Basic .....' } })
+      );
+
+      const authenticationResult = await authenticator.login(request, {
+        provider: 'basic',
+        value: {},
+      });
+      expect(authenticationResult.succeeded()).toBe(true);
+      expect(authenticationResult.user).toEqual(user);
+      expect(authenticationResult.authHeaders).toEqual({ authorization: 'Basic .....' });
+    });
+
+    it('creates session whenever authentication provider returns state for system API requests', async () => {
+      const user = mockAuthenticatedUser();
+      const request = requestFixture();
+      const authorization = `Basic ${Buffer.from('foo:bar').toString('base64')}`;
+
+      mockOptions.isSystemAPIRequest.mockReturnValue(true);
+      mockBasicAuthenticationProvider.login.mockResolvedValue(
+        AuthenticationResult.succeeded(user, { state: { authorization } })
+      );
+
+      const systemAPIAuthenticationResult = await authenticator.login(request, {
+        provider: 'basic',
+        value: {},
+      });
+      expect(systemAPIAuthenticationResult.succeeded()).toBe(true);
+      expect(systemAPIAuthenticationResult.user).toEqual(user);
+
+      expect(mockSessionStorage.set).toHaveBeenCalledTimes(1);
+      expect(mockSessionStorage.set).toHaveBeenCalledWith({
+        expires: null,
+        state: { authorization },
+        provider: 'basic',
+      });
+    });
+
+    it('creates session whenever authentication provider returns state for non-system API requests', async () => {
```

Comment:
Hmm, I'm not sure I'm following. Can you please elaborate on that a bit? Why can't this (I assume `this == "login" creates session whenever authentication provider returns state for non-system API requests`) happen?

Reply:
Apologies for not being clear. It's my understanding that the login method is only used by the [login route](https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/security/server/routes/api/v1/authenticate.js#L16) [saml route](https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/security/server/routes/api/v1/authenticate.js#L49) and the [oidc route](https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/security/server/routes/api/v1/authenticate.js#L77) which really shouldn't ever be "system api requests" (done in the background by the system). Additionally, shouldn't we be starting the session always in these scenarios instead of worrying about extending the existing one?

Reply:
Thanks for clarifying! I think I'm got the idea now.

> It's my understanding that the login method is only used by the login route, saml route and the oidc route which really shouldn't ever be "system api requests" (done in the background by the system).

Yep, that's correct or any other "auth entry point" route we'll define in the future.

> Additionally, shouldn't we be starting the session always in these scenarios instead of worrying about extending the existing one?

Yes, I believe so. Let me experiment with this and see what I can come up with.

Reply:
I spent some time thinking about this and I can't find enough justification for a dedicated `LoginResult` _yet_, it feels like `LoginResult` may need all the properties and methods that `AuthenticationResult` has. There are two arguable things though: `shouldClearState` and `notHandled`. I tried to explain in code comment below why we may need `shouldClearState` in `login` (clear existing state without creating a new one). Basically I made this use case up, let me know if it doesn't make any sense to you? And we use `notHandled` as a way to indicate that provider doesn't support `login`.

Having said I definitely agree that using `updateSessionValue` in `login` is not necessary and very confusing. Here is what we can use instead. How do you feel about it?

```typescript
// There are two possible cases when we'd want to clear existing state:
// 1. If provider owned the state (e.g. intermediate state used for multi step login), but failed
// to login, that likely means that state is not valid anymore and we should clear it.
// 2. Also provider can specifically ask to clear state by setting it to `null` even if
// authentication attempt didn't fail (e.g. custom realm could "pin" client/request identity to
// a server-side only session established during multi step login that relied on intermediate
// client-side state that we don't need anymore).
if (
  authenticationResult.shouldClearState() ||
  (authenticationResult.failed() && getErrorStatusCode(authenticationResult.error) === 401)
) {
  sessionStorage.clear();
} else if (authenticationResult.shouldUpdateState()) {
  sessionStorage.set({
    state: authenticationResult.state,
    provider: attempt.provider,
    expires: this.ttl && Date.now() + this.ttl,
  });
}
```

Reply:
That sounds good to me! Thanks for giving this so much thought.